File: clojure/triangle/src/triangle.clj
(ns triangle)

(defn- valid-triangle? [a b c]
  (and (> (+ a b) c) (> (+ a c) b) (> (+ b c) a)))

(defn- isosceles? [a b c]
  (or (= a b) (= a c) (= b c)))

(defn type [a b c]
  (cond (and (= a b) (= a c)) :equilateral
        (and (valid-triangle? a b c) (isosceles? a b c)) :isosceles
        (valid-triangle? a b c) :scalene
        :else :illogical))

; ======================================================================
; Inspire by Official Solution: sorted before compare. count distinct.
; ======================================================================

(defn type [a b c]
  (let [sorted (sort > [a b c])]
    (if (>= (first sorted) (apply + (rest sorted))) :illogical
        (case (count (distinct sorted))
          1 :equilateral
          2 :isosceles
          3 :scalene))))

File: javascript/triangle/triangle.js
function Triangle (a, b, c) {
    'use strict';
    this.borders = [a, b, c].sort();
}

function isInvalid(borders) {
    'use strict';
    function negative (currentValue) {
        return currentValue <= 0;
    }

    if (borders.some(negative)) {
        return true;
    }

    if (borders[0] + borders[1] <= borders[2]) {
        return true;
    }

    return false;
}

Triangle.prototype.kind = function () {
    'use strict';
    var b = this.borders;

    if (isInvalid(b)) {
        throw new TypeError('illegal');
    } else if (b[0] === b[1] && b[0] === b[2]) {
        return 'equilateral';
    } else if (b[0] === b[1] || b[1] === b[2]) {
        return 'isosceles';
    } else {
        return 'scalene';
    }
};

module.exports = Triangle;

